### DISCUSSION TITLE
how to deep marshal the block to JSON in fabric-protos-go-apiv2

### DISCUSSION THREAD

**Discussion by 1gezhanghao** [hyperledger/fabric-gateway#578] [Q&A]
In  [fabric-protos-go v1](github.com/hyperledger/fabric-protos-go/),we can use github.com/hyperledger/fabric-config/protolator to DeepMarshalJSON a block
```
protolator.DeepMarshalJSON(&buffer, block)
```

In [fabric-protos-go-apiv2](github.com/hyperledger/fabric-protos-go-apiv2) , how to Deep Marshal a block to json?

**Comment by bestbeforetoday** [comment]
Protolator package looks to be implemented in terms of the deprecated v1 Golang protobuf API, whereas fabric-protos-apiv2 (as the name suggests) uses the v2 Golang protobuf API. You _might_ be able to convert the v2 to a v1 message using [proto.MessageV1](https://pkg.go.dev/github.com/golang/protobuf/proto) and then pass that to the protolator functions. I haven’t personally tried this and it might be that mixing the two versions of the Fabric protobuf messages in the same project will cause [protobuf namespace conflicts](https://protobuf.dev/reference/go/faq/#namespace-conflict).

The closest thing to an out-of-the box solution is the protobuf [protojson](https://pkg.go.dev/google.golang.org/protobuf/encoding/protojson) package, but this will of course not unpack nested serialised protobuf messages, which is exactly the issue with Fabric’s complicated protobuf structure that protolator aims to address.

Perhaps the ideal solution would be for a v2 protolator package to implemented using the updated protobuf API and Fabric protobuf bindings.

**Comment by 1gezhanghao** [comment]
yes, we cann't use  two versions of the Fabric protobuf messages in the same project that will cause [protobuf namespace conflicts](https://protobuf.dev/reference/go/faq/#namespace-conflict).
and also i tried [protojson](https://pkg.go.dev/google.golang.org/protobuf/encoding/protojson) ,it cann't deal with nested data struct and some data format become weird
that's why i ask for help.

the old protolator util project sames deprecated https://github.com/hyperledger/fabric-config/issues/53

Its a good  idea to implements a v2 protolator package using the updated protobuf API and Fabric protobuf bindings.
